[wasm] [debugger] First version of multithreaded debugging

thaystg opened the pull request:
https://user-images.githubusercontent.com/4503299/187486582-7b63330a-72f2-49fb-8efd-0e0a300f0a0c.mp4

Build command:
`.\build.cmd mono+libs -os browser /p:MonoWasmBuildVariant=multithread`

Run debugger-tests command:
`.\dotnet.cmd test src/mono/wasm/debugger/DebuggerTestSuite --filter DebuggerTests.MiscTests.TestDebugUsingMultiThreadedRuntime -e RuntimeConfiguration=Debug -e Configuration=Debug -e DebuggerHost=chrome -e WASM_TESTS_USING_VARIANT=multithreaded -e WasmEnableThreads=true`

thaystg committed:
First version of multithreaded debugging.

thaystg committed:
Merge branch 'main' into thays_support_multithreaded_debugging

thaystg committed:
Revert package-lock.json

thaystg committed:
New line at package-lock.json

thaystg committed:
Fix not used variable.

thaystg committed:
Fix debugger on firefox.

thaystg committed:
Merge branch 'thays_support_multithreaded_debugging' of github.com:thaystg/runtime into thays_support_multithreaded_debugging

thaystg commented:
/azp run runtime-wasm

thaystg committed:
Rewrite code to avoid duplicated code.

thaystg committed:
Fix where mono_init_debugger_agent_common is called.

thaystg committed:
Remove whitespace.

radical commented:
The Wasm.Build.Tests failures are https://github.com/dotnet/runtime/issues/74944 .

lambdageek commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:
```diff
@@ -433,7 +440,7 @@ public ExecutionContext(MonoSDBHelper sdbAgent, int id, object auxData, PauseOnE
         internal DebugStore store;
         internal MonoSDBHelper SdbAgent { get; init; }
-        public TaskCompletionSource<DebugStore> Source { get; } = new TaskCompletionSource<DebugStore>();
+        public TaskCompletionSource<DebugStore> Source { get; set; } = new TaskCompletionSource<DebugStore>();
```
nit: could be `{get; private init;}`

and then

```csharp
var ret = new ExecutionContext(SdbAgent.Clone(sessionId), Id, AuxData, PauseOnExceptions) {
    Source = Source }
```

lambdageek reviewed (commented):
C code and TS code lgtm.

I'd like someone more familiar with the debug proxy to review that part

radical commented:
needs to merge main

thaystg committed:
Merge branch 'main' into thays_support_multithreaded_debugging

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/Firefox/FirefoxMonoProxy.cs:
```diff
@@ -710,7 +720,7 @@ private async Task<bool> SendPauseToBrowser(SessionId sessionId, JObject args, C
             return false;
 
-        byte[] newBytes = Convert.FromBase64String(res.Value?["result"]?["value"]?["value"]?.Value<string>());
+        byte[] newBytes = Convert.FromBase64String(res.Value?["result"]?[0]?["value"]?["value"]?.Value<string>());
```
Am I correct in understanding that this is reading the first argument of the js function? Can we read it by name instead, so we are sure of what we are reading?

thaystg replied:
We cannot read by name, because when it's in release mode we don't know de parameter name anymore.

radical replied:
ok! Then we can check `res.Value?["result"]?` has any elements, if not then log a relevant error.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1053,15 +1069,24 @@ internal virtual void SaveLastDebuggerAgentBufferReceivedToContext(SessionId ses
-            if (!res.IsOk)
+            var argsNew = JObject.FromObject(new
+            {
+                objectId = args?["callFrames"]?[0]?["scopeChain"]?[0]?["object"]?["objectId"]?.Value<string>(),
```
Same comment here, can we read the `data` parameter of the js function by name?

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1053,15 +1069,24 @@ internal virtual void SaveLastDebuggerAgentBufferReceivedToContext(SessionId ses
+        internal async Task<bool> OnReceiveDebuggerAgentEvent(SessionId sessionId, JObject args, Result debuggerAgentBuffer, CancellationToken token)
+        {
+            SaveLastDebuggerAgentBufferReceivedToContext(sessionId, debuggerAgentBuffer);
+            if (!debuggerAgentBuffer.IsOk)
```
This should be logged. It would be an unusual error, IIUC.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1053,15 +1069,24 @@ internal virtual void SaveLastDebuggerAgentBufferReceivedToContext(SessionId ses
+
+        internal async Task<bool> OnReceiveDebuggerAgentEvent(SessionId sessionId, JObject args, Result debuggerAgentBuffer, CancellationToken token)
+        {
+            SaveLastDebuggerAgentBufferReceivedToContext(sessionId, debuggerAgentBuffer);
```
I see that this has an implementation only for firefox, but why does it need to save it?

thaystg replied:
Because from firefox we receives 2 different messages, one with the pause and another asking for the callstack.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1053,15 +1069,24 @@ internal virtual void SaveLastDebuggerAgentBufferReceivedToContext(SessionId ses
         }
 
-        internal async Task<bool> OnReceiveDebuggerAgentEvent(SessionId sessionId, JObject args, CancellationToken token)
+        internal async Task<Result> GetLastDebuggerAgentBuffer(SessionId sessionId, JObject args, CancellationToken token)
```
Will the buffer be relevant once `OnReceiveDebuggerAgentEvent` is done? If not, then this method should be merged with that.

thaystg replied:
For firefox yes.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1286,7 +1311,7 @@ private async Task<bool> OnAssemblyLoadedJSEvent(SessionId sessionId, JObject ev
                 foreach (var source in store.Add(sessionId, assembly_name, assembly_data, pdb_data, token))
                 {
-                    await OnSourceFileAdded(sessionId, source, context, token);
+                    await OnSourceFileAdded(sessionId, source, context, false, token);
```
named params, because it isn't obvious what the `false` means.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -807,6 +807,14 @@ public MonoSDBHelper(MonoProxy proxy, ILogger logger, SessionId sessionId)
             ValueCreator = new(this, logger);
             ResetStore(null);
         }
+        public MonoSDBHelper Clone(SessionId sessionId)
+        {
+            var ret = new MonoSDBHelper(this.proxy, this.logger, sessionId);
+            ret.VmMajorVersion = this.VmMajorVersion;
+            ret.VmMinorVersion = this.VmMinorVersion;
+            ret.store = this.store;
+            return ret;
+        }
```
```suggestion

        public MonoSDBHelper Clone(SessionId sessionId)
            => new MonoSDBHelper(proxy, logger, sessionId)
            {
                VmMajorVersion = VmMajorVersion,
                VmMinorVersion = VmMinorVersion,
                store = store,
            };
```

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1462,12 +1487,13 @@ private async Task<Breakpoint> SetMonoBreakpoint(SessionId sessionId, string req
         }
 
-        internal virtual async Task OnSourceFileAdded(SessionId sessionId, SourceFile source, ExecutionContext context, CancellationToken token)
+        internal virtual async Task OnSourceFileAdded(SessionId sessionId, SourceFile source, ExecutionContext context, bool ignoreBreakpoint, CancellationToken token)
```
Maybe `bool resolveBreakpoints = true`

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -243,6 +246,17 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, JObject par
+            var store = await LoadStore(sessionId, true, token);
+            foreach (var source in store.AllSources())
+            {
+                await OnSourceFileAdded(sessionId, source, contexts[sessionId], true, token);
```
So, we need to send out `scriptParsed` events for the new sessionId? If we can identify when a context gets created for a non-main thread of the app being debugged, then this wouldn't be needed, I think.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:
```diff
@@ -408,7 +408,14 @@ public ExecutionContext(MonoSDBHelper sdbAgent, int id, object auxData, PauseOnE
             PauseOnExceptions = pauseOnExceptions;
         }
-
+        public ExecutionContext Clone(SessionId sessionId)
+        {
+            var ret = new ExecutionContext(SdbAgent.Clone(sessionId), Id, AuxData, PauseOnExceptions);
+            ret.ready = ready;
+            ret.store = store;
+            ret.Source = Source;
+            return ret;
+        }
```
```suggestion
        public ExecutionContext Clone(SessionId sessionId)
           => new ExecutionContext(SdbAgent.Clone(sessionId), Id, AuxData, PauseOnExceptions)
            {
                ready = ready,
                store = store,
                Source = Source
            };
```

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

radical committed:
[wasm] Debugger tests: support running with multithreaded runtime

radical committed:
Add runtime-wasm-dbgtests pipeline with debugger tests running on a multi-threaded runtime

radical committed:
Add multi-threaded debugger tests to runtime-wasm

radical committed:
fix yml

radical commented:
/azp run runtime-wasm-non-libtests

radical committed:
Always run the new tests when the pipeline is invoked manually

radical commented:
/azp run runtime-wasm-non-libtests

radical committed:
Pass through extra build args for wasm debugger tests

radical commented:
/azp run runtime-wasm-non-libtests

radical commented:
[Log](https://helix.dot.net/api/2019-06-17/jobs/ebed2784-7967-493b-9645-ca6090e2de0e/workitems/chrome-DebuggerTests/console):

- 17 failures
- there are some timeouts
- Others look like async related tests

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -169,9 +171,11 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, JObject par
-                        if (!contexts.ContainsKey(sessionId))
-                            return false;
+                        // Don't process events from sessions we aren't tracking unless they have asyncStackTraceId
+                        if (!contexts.ContainsKey(sessionId) && args["asyncStackTraceId"] != null)
```
Fixed this, now we create the execution context when we receive the attach to the worker.

thaystg committed:
Addressing @radical comments.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1503,7 +1542,7 @@ internal virtual async Task<DebugStore> LoadStore(SessionId sessionId, bool tryU
                     await foreach (SourceFile source in context.store.Load(sessionId, loaded_files, context, useDebuggerProtocol, token))
                     {
-                        await OnSourceFileAdded(sessionId, source, context, token);
+                        await OnSourceFileAdded(sessionId, source, context, token, false);
```
Should this pass `resolveBreakpoints: true`? And `false` for the worker case?

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:
```diff
@@ -432,8 +449,8 @@ public ExecutionContext(MonoSDBHelper sdbAgent, int id, object auxData, PauseOnE
         internal DebugStore store;
-        internal MonoSDBHelper SdbAgent { get; init; }
-        public TaskCompletionSource<DebugStore> Source { get; } = new TaskCompletionSource<DebugStore>();
+        internal MonoSDBHelper SdbAgent { get; set; }
```
```suggestion
        internal MonoSDBHelper SdbAgent { get; private set; }
```

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:
```diff
@@ -419,6 +434,8 @@ public ExecutionContext(MonoSDBHelper sdbAgent, int id, object auxData, PauseOnE
         public bool IsResumedAfterBp { get; set; }
         public int ThreadId { get; set; }
         public int Id { get; set; }
+        public ExecutionContext ParentContext { get; set; }
+        public SessionId SessionId { get; set; }
```
```suggestion
        public ExecutionContext ParentContext { get; private set; }
        public SessionId SessionId { get; private set; }
```

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
addressing radical comments

thaystg committed:
Merge branch 'dotnet:main' into thays_support_multithreaded_debugging

thaystg committed:
Fixing tests failures and adding a schema to run a test that will only run in a multithreaded environment.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -36,6 +36,11 @@ public static WasmHost RunningOn
             => WasmHost.Firefox;
+#endif
+#if WASM_ENABLE_THREADS
+        public static bool WasmEnableThreads = true;
```
I would suggest an environment variable like `WASM_TESTS_USING_VARIANT=..` which have a default empty value, and can optionally be `=multithreaded` to run with the mt variant of the runtime.

thaystg committed:
Adding support for run debugger-tests in a multithreaded runtime.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -37,6 +37,18 @@ public static WasmHost RunningOn
             => WasmHost.Firefox;
 #endif
+
+        public static bool WasmEnableThreads 
+        {
+            get
+            {
+                var envVar = Environment.GetEnvironmentVariable("WASM_TESTS_USING_VARIANT");
+                if (envVar == "multithreaded")
+                    return true;
+                return false;
+            }
+        }
```
Instead of this, add:
```csharp
public static readonly string? WasmTestsUsingVariant = Environment.GetEnvironmentVariable("WASM_TESTS_USING_VARIANT");
```
.. to `EnvironmentVariables.cs`.

```suggestion
        public static bool WasmEnableThreads => EnvironmentVariables.WasmTestsUsingVariant == "multithreaded";
```

thaystg replied:
done.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -133,15 +145,17 @@ public virtual async Task InitializeAsync()
+                    getInitCmdFn("Runtime.enable", null),
+                    getInitCmdFn("Debugger.enable", null),
+                    getInitCmdFn("Runtime.runIfWaitingForDebugger", null),
+                    getInitCmdFn("Debugger.setAsyncCallStackDepth", JObject.FromObject(new { maxDepth = 32})),
+                    getInitCmdFn("Target.setAutoAttach", JObject.FromObject(new { autoAttach = true, waitForDebuggerOnStart = true, flatten = true}))
```
Are these relevant for single threaded runtime too?

radical replied:
```suggestion
                    getInitCmdFn("Target.setAutoAttach", JObject.FromObject(new { autoAttach = true, waitForDebuggerOnStart = true, flatten = true }))
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -133,15 +145,17 @@ public virtual async Task InitializeAsync()
+                    getInitCmdFn("Runtime.enable", null),
+                    getInitCmdFn("Debugger.enable", null),
+                    getInitCmdFn("Runtime.runIfWaitingForDebugger", null),
+                    getInitCmdFn("Debugger.setAsyncCallStackDepth", JObject.FromObject(new { maxDepth = 32})),
```
What will the `32` mean in practice? The documentation doesn't make it clear for me - https://chromedevtools.github.io/devtools-protocol/tot/Debugger/#method-setAsyncCallStackDepth

thaystg replied:
I copied the value that I see in the logs when I debug from VSCode, so I don't know exactly what it means.

radical replied:
to be figured out later then :)

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -133,15 +145,17 @@ public virtual async Task InitializeAsync()
+                    getInitCmdFn("Runtime.enable", null),
+                    getInitCmdFn("Debugger.enable", null),
+                    getInitCmdFn("Runtime.runIfWaitingForDebugger", null),
+                    getInitCmdFn("Debugger.setAsyncCallStackDepth", JObject.FromObject(new { maxDepth = 32})),
```
```suggestion
                    getInitCmdFn("Debugger.setAsyncCallStackDepth", JObject.FromObject(new { maxDepth = 32 })),
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -199,6 +199,16 @@ async Task OnMessage(string method, JObject args, CancellationToken token)
+                case "Target.attachedToTarget":
+                {
+                    var sessionId = new SessionId(args["sessionId"]?.Value<string>());
+                    await Client.SendCommand(sessionId, "Profiler.enable", null, token);
```
We should use the code for handling these init commands in `OpenSessionAsync` here also. It could be extracted to a new method. It can be done in a subsequent PR too.

thaystg committed:
Fix running debugger tests for multithreaded runtime, passing sessionId where it's necessary.

thaystg commented:
/azp run runtime-wasm-non-libtests

thaystg committed:
Fix CI.

thaystg commented:
/azp run runtime-wasm-non-libtests

radical commented:
re:sessionId, for multithreaded runtime case, we'll have 1 or more ids - 1 main sessionId, and the others being for the other threads. When would the non-main ones be useful in the tests, and how would you expect to use them?
I'm just wondering if the `Inspector` can have a list of all the sessionIds, and by default we use the main id everywhere, but allow passing a different one in relevant test APIs. How does that sound?

thaystg committed:
Addressing @radical comments
Adding a test case.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/MiscTests.cs:
```diff
@@ -1129,5 +1129,38 @@ public async Task InspectReadOnlySpan()
+            var bp = await SetBreakpointInMethod("debugger-test.dll", "MultiThreadedTest", "Write", 2);
+            var expression = $"{{ invoke_static_method('[debugger-test] MultiThreadedTest:Run'); }}";
+
+            await EvaluateAndCheck(
+                "window.setTimeout(function() {" + expression + "; }, 1);",
+                "dotnet://debugger-test.dll/debugger-test.cs", 1529, 8,
+                "MultiThreadedTest.Write",
+                wait_for_event_fn: async (pause_location) =>
+                {
+                    var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());
+                    Assert.Equal(locals[1]["value"]["type"], "number");
+                    Assert.Equal(locals[1]["name"], "currentThread");
+                    await StepAndCheck(StepKind.Resume, "dotnet://debugger-test.dll/debugger-test.cs", 1529, 8, "MultiThreadedTest.Write",
+                    wait_for_event_fn: async (pause_location) =>
+                    {
+                        var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());
+                        Assert.Equal(locals[1]["value"]["type"], "number");
+                        Assert.Equal(locals[1]["name"], "currentThread");
+                        await StepAndCheck(StepKind.Resume, "dotnet://debugger-test.dll/debugger-test.cs", 1529, 8, "MultiThreadedTest.Write",
+                        wait_for_event_fn: async (pause_location) =>
+                        {
+                            var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());
+                            Assert.Equal(locals[1]["value"]["type"], "number");
+                            Assert.Equal(locals[1]["name"], "currentThread");
+                        });
+                    });
+                }
+            );
```
The nesting can be avoided, and that would mean better traces when something fails. I didn't try to compile this:

```suggestion
            var pause_location = await EvaluateAndCheck(
                "window.setTimeout(function() {" + expression + "; }, 1);",
                "dotnet://debugger-test.dll/debugger-test.cs", 1529, 8,
                "MultiThreadedTest.Write");

            var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());
            Assert.Equal(locals[1]["value"]["type"], "number");
            Assert.Equal(locals[1]["name"], "currentThread");

            pause_location = await StepAndCheck(StepKind.Resume, "dotnet://debugger-test.dll/debugger-test.cs", 1529, 8, "MultiThreadedTest.Write");
            var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());
            Assert.Equal(locals[1]["value"]["type"], "number");
            Assert.Equal(locals[1]["name"], "currentThread");

            pause_location = await StepAndCheck(StepKind.Resume, "dotnet://debugger-test.dll/debugger-test.cs", 1529, 8, "MultiThreadedTest.Write");
            var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());
            Assert.Equal(locals[1]["value"]["type"], "number");
            Assert.Equal(locals[1]["name"], "currentThread");
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -290,7 +305,7 @@ public async Task WaitForScriptParsedEventsAsync(params string[] paths)
 
             Assert.Equal(bp.Value["breakpointId"]?.ToString(), pause_location["hitBreakpoints"]?[0]?.Value<string>());
-
+        
```
```suggestion

```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs:
```diff
@@ -219,23 +219,25 @@ public async Task ConditionalBreakpoint2(string function_to_call, string method_
+        [InlineData("c == 15", 79, 3, 79, 11, "dotnet.worker.js")]
+        [InlineData("c == 17", 79, 3, 80, 11, "debugger-driver.html")]
+        [InlineData("g == 17", 79, 3, 80, 11, "debugger-driver.html")]
+        [InlineData("true", 79, 3, 79, 11, "dotnet.worker.js")]
```
What does the trace look like for this? Just wondering why this is in `dotnet.worker.js`.

thaystg replied:
Thanks for the question, there was an issue, we are receiving the same scriptId, so we are replacing in the dictionary wrongly.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -229,6 +243,22 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, JObject par
+            }
+            if (contexts.ContainsKey(workerSessionId))
+            {
+                logger.LogDebug($"Worker sessionId already exists - {originSessionId}");
```
This shouldn't happen, right? Since, multithreaded debugging is WIP right now, it might be more useful to throw an exception here.
Also, we should handle `Runtime.executionContextCreated`, and `Runtime.executionContextsCleared` to remove them from `contexts`. And `contexts` might need to handle being modified in parallel. This can be done in a follow up PR though.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -81,12 +81,20 @@ public Task<JObject> WaitFor(string what)
             {
                 if (tcs.Task.IsCompleted)
                 {
+                    Client.CurrentSessionId = new SessionId(tcs.Task.Result["sessionId"]?.Value<string>());
```
I'm not sure I understand this. The client is being switched to the session id based on the last message we received? Why is that?

thaystg replied:
When we receive a pause and will process the pause, we want to run the next commands like get properties, evaluate commands, in the session that we are paused.

radical replied:
Maybe it would make sense to keep track of the "main thread"'s sessionId, and the related sessions. And also have a `PausedOnSessionId`, which we set/clear on pause/resume.

For future work, it would be useful to know - do we get contextCreated events per worker, or per managed thread? And what does it mean to evaluate a managed expression on a main vs worker context(sessionId) ? I guess, we'll find out more as we try this out more.

thaystg committed:
Update src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Update src/mono/wasm/debugger/DebuggerTestSuite/MiscTests.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Update src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Dictionary with the scriptId also uses sessionId.

thaystg committed:
Addressing @radical review.

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Avoiding getting this error: Cannot transition thread 0x2a15360 from STATE_BLOCKING with DO_BLOCKING.
In the transport_send we don't save the thread context, we save it before the send function.

thaystg committed:
Addressing @radical comments.

thaystg commented:
/azp run runtime-wasm-non-libtests

thaystg committed:
Using more threads in unit test.

thaystg commented:
/azp run runtime-wasm-non-libtests

radical commented on src/mono/mono/component/mini-wasm-debugger.c:
```diff
@@ -382,7 +382,7 @@ mono_wasm_send_dbg_command_with_parms (int id, MdbgProtCommandSet command_set, i
 	}
 	MdbgProtBuffer bufWithParms;
-	buffer_init (&bufWithParms, 128);
+	m_dbgprot_buffer_init (&bufWithParms, 128);
```
`debugger-engine.h` has `#define buffer_init m_dbgprot_buffer_init`. Is that not applicable here?

thaystg replied:
yeah, I just used the full name because we also use m_dbgprot_buffer_add_int and m_dbgprot_buffer_add_byte_array.

radical reviewed (approved):
This looks great! The remaining comments can be addressed in follow up PRs. 👍 🥳

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

jakobbotsch commented:
Sorry about the pgostress pipeline clutter... there were some trigger issues with a newly created pipeline. It should be fixed now but I'm afraid it will stay under the checks.

thaystg committed:
Addressing @radical comments, and trying to fix ci.

thaystg commented:
/azp run runtime-wasm-non-libtests

thaystg committed:
Merge remote-tracking branch 'origin/main' into thays_support_multithreaded_debugging

# Conflicts:
#	src/mono/wasm/debugger/BrowserDebugProxy/Firefox/FirefoxMonoProxy.cs
#	src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs
#	src/mono/wasm/debugger/DebuggerTestSuite/EnvironmentVariables.cs
#	src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs
#	src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj

thaystg committed:
Removing unnecessary changes.

thaystg committed:
Export function used on mini-wasm-debugger.

lambdageek commented:
@thaystg  I added a tracking issue for threaded debugger work in case we find anything we want to do in follow-up PRs https://github.com/dotnet/runtime/issues/81282

thaystg committed:
Fixing line number.

thaystg commented:
/azp run runtime-wasm

thaystg committed:
Merge branch 'dotnet:main' into thays_support_multithreaded_debugging

thaystg commented:
/azp run runtime-wasm

thaystg committed:
Fix run tests on release.

thaystg commented:
/azp run runtime-wasm

thaystg committed:
fix compilation for multithread runtime

thaystg committed:
trying to fix multithread debugger tests on ci

thaystg commented:
/azp run runtime-wasm

thaystg committed:
trying to fix debugger tests on ci

thaystg committed:
Merge branch 'dotnet:main' into thays_support_multithreaded_debugging

thaystg commented:
/azp run runtime-wasm-non-libtests

thaystg committed:
disabling tests on multithreaded runtime

thaystg commented:
/azp run runtime-wasm-non-libtests

thaystg commented:
@radical @lambdageek Could you please do a final review so I can merge?

lambdageek reviewed (approved):
C and TypeScript lgtm.  I tried to follow BrowserDebugProxy, but I'm not sure if I completely understand it

radical commented on eng/pipelines/runtime-wasm-dbgtests.yml:
```diff
@@ -0,0 +1,41 @@
+#
+# Evaluate paths
+#
+- template: /eng/pipelines/common/evaluate-default-paths.yml
```
This is fine for now. But in a follow up we should change this to use these from runtime-extra-platforms-wasm.yml with a new `debuggerTestsOnly` parameter.

radical commented on eng/pipelines/extra-platforms/runtime-extra-platforms-wasm.yml:
```diff
@@ -213,6 +213,17 @@ jobs:
+        - Browser_wasm_win
+      extraBuildArgs: /p:MonoWasmBuildVariant=multithread /p:WasmEnableThreads=true
+      nameSuffix: DebuggerTests_MultiThreaded
+      alwaysRun: true # always run when invoked manually
```
```suggestion
      alwaysRun: ${{ parameters.isWasmOnlyBuild }}
```

This will run whenever the pipeline is invoked manually.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -81,12 +81,20 @@ public Task<JObject> WaitFor(string what)
 
                 throw new Exception($"Invalid internal state, waiting for {what} while another wait is already setup");
             }
+            else if (nextNotifications.TryDequeue(out var notification))
```
What if the next notification in the queue is for a different `string what`?

thaystg replied:
I'll push some changes and it will be throwing an exception if it's a different `string what`

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -194,14 +206,28 @@ void FailAllWaiters(Exception? exception = null)
+                case "Target.attachedToTarget":
+                {
+                    var sessionIdNewTarget = new SessionId(args["sessionId"]?.Value<string>());
+                    await Client.SendCommand(sessionIdNewTarget, "Profiler.enable", null, token);
+                    await Client.SendCommand(sessionIdNewTarget, "Runtime.enable", null, token);
+                    await Client.SendCommand(sessionIdNewTarget, "Debugger.enable", null, token);
+                    await Client.SendCommand(sessionIdNewTarget, "Runtime.runIfWaitingForDebugger", null, token);
+                    await Client.SendCommand(sessionIdNewTarget, "Debugger.setAsyncCallStackDepth", JObject.FromObject(new { maxDepth = 32}), token);
```
For a follow up PR - the commands here, and in `DebuggerTestBase.InitializeAsync` should use the same base set of commands. And each case can add any additional ones as needed. Like the ones here seem to be the same.

thaystg committed:
Update eng/pipelines/extra-platforms/runtime-extra-platforms-wasm.yml

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Merge branch 'dotnet:main' into thays_support_multithreaded_debugging

thaystg committed:
Throwing an exception if the "what" is not the one that is being get from the nextNotificationQueue.

thaystg commented:
/azp run runtime-wasm-non-libtests

thaystg commented:
/azp run runtime-wasm-non-libtests
